src/snowflake/snowpark/mock/plan.py

```diff
@@ -808,11 +818,275 @@ def outer_join(base_df):
+        join_result.sf_types = cartesian_product.sf_types
+
+        # TODO [GA]: # ERROR_ON_NONDETERMINISTIC_MERGE is by default True, raise error if
+        # (1) A target row is selected to be updated with multiple values OR
+        # (2) A target row is selected to be both updated and deleted
+
+        inserted_rows = []
+        inserted_row_idx = set()  # source_row_id
+        deleted_row_idx = set()
+        updated_row_idx = set()
+        for clause in source_plan.clauses:
+            if isinstance(clause, UpdateMergeExpression):
+                # Select rows to update
+                if clause.condition:
+                    condition = calculate_expression(
+                        clause.condition, join_result, analyzer, expr_to_alias
+                    )
+                    rows_to_update = join_result[condition]
+                else:
+                    rows_to_update = join_result
+
+                rows_to_update = rows_to_update[
+                    ~rows_to_update[ROW_ID]
+                    .isin(updated_row_idx.union(deleted_row_idx))
+                    .values
+                ]
+
+                # Update rows in place
+                for attr, new_expr in clause.assignments.items():
+                    column_name = analyzer.analyze(attr, expr_to_alias)
+                    target_index = target.loc[rows_to_update[ROW_ID]].index
+                    new_val = calculate_expression(
+                        new_expr, rows_to_update, analyzer, expr_to_alias
+                    )
+                    new_val.index = target_index
+                    target.loc[rows_to_update[ROW_ID], column_name] = new_val
+
+                # Update updated row id set
+                for _, row in rows_to_update.iterrows():
+                    updated_row_idx.add(row[ROW_ID])
+
+            elif isinstance(clause, DeleteMergeExpression):
+                # Select rows to delete
+                if clause.condition:
+                    condition = calculate_expression(
+                        clause.condition, join_result, analyzer, expr_to_alias
+                    )
+                    intermediate = join_result[condition]
+                else:
+                    intermediate = join_result
+
+                matched = target.apply(tuple, 1).isin(
+                    intermediate[target.columns].apply(tuple, 1)
+                )
+                matched.sf_type = ColumnType(BooleanType(), True)
+
+                # Update deleted row id set
+                for _, row in target[matched].iterrows():
+                    deleted_row_idx.add(row[ROW_ID])
+
+                # Delete rows in place
+                target = target[~matched]
+
+            elif isinstance(clause, InsertMergeExpression):
+                # calculate unmatched rows in the source
+                matched = source.apply(tuple, 1).isin(
+                    join_result[source.columns].apply(tuple, 1)
+                )
```

Comment:
knowledge question: how this works, looking at the code it feels like matched == source?

Reply:
This is getting the rows in source table that also exist in the join result (`cartesian_product[join_condition]`). We use `apply(tuple, 1)` to flatten a row into a tuple of values and use `is_in` to get a column of boolean values that represent whether each row in `source` exists in the join result.